autofs-5.0.7 - probe each nfs version in turn for singleton mounts

From: Ian Kent <raven@themaw.net>

Original patch + subsequent fix patch.

---
 CHANGELOG            |    1 +
 include/replicated.h |    2 ++
 modules/mount_nfs.c  |   35 ++++++++++++++++++++++++++++++++++-
 modules/replicated.c |    8 ++++----
 4 files changed, 41 insertions(+), 5 deletions(-)

--- autofs-5.0.5.orig/CHANGELOG
+++ autofs-5.0.5/CHANGELOG
@@ -127,6 +127,7 @@
 - fix master map mount options matching.
 - fix master map bogus keywork match.
 - dont probe rdma mounts.
+- probe each nfs version in turn for singleton mounts.
 
 03/09/2009 autofs-5.0.5
 -----------------------
--- autofs-5.0.5.orig/include/replicated.h
+++ autofs-5.0.5/include/replicated.h
@@ -68,6 +68,8 @@ struct host {
 };
 
 void seed_random(void);
+struct host *new_host(const char *, struct sockaddr *, size_t,
+		      unsigned int, unsigned int, unsigned int);
 void free_host_list(struct host **);
 int parse_location(unsigned, struct host **, const char *, unsigned int);
 int prune_host_list(unsigned, struct host **, unsigned int, int);
--- autofs-5.0.5.orig/modules/mount_nfs.c
+++ autofs-5.0.5/modules/mount_nfs.c
@@ -186,9 +186,42 @@ int mount_mount(struct autofs_point *ap,
 	 * We can't probe protocol rdma so leave it to mount.nfs(8)
 	 * and and suffer the delay if a server isn't available.
 	 */
-	if (!rdma)
+	if (rdma)
+		goto dont_probe;
+
+	/*
+	 * If this is a singleton mount, and NFSv4 only hasn't been asked
+	 * for, and the default NFS protocol is set to v4 in the autofs
+	 * configuration only probe NFSv4 and let mount.nfs(8) do fallback
+	 * to NFSv3 (if it can). If the NFSv4 probe fails then probe as
+	 * normal.
+	 */
+	if ((hosts && !hosts->next) &&
+	    mount_default_proto == 4 &&
+	    vers & NFS_VERS_MASK != 0 &&
+	    vers & NFS4_VERS_MASK != 0) {
+		unsigned int v4_probe_ok = 0;
+		struct host *tmp = new_host(hosts->name,
+					    hosts->addr, hosts->addr_len,
+					    hosts->proximity,
+					    hosts->weight, hosts->options);
+		if (tmp) {
+			tmp->rr = hosts->rr;
+			prune_host_list(ap->logopt, &tmp,
+					NFS4_VERS_MASK|TCP_SUPPORTED, port);
+			/* If probe succeeds just try the mount with host in hosts */
+			if (tmp) {
+				v4_probe_ok = 1;
+				free_host_list(&tmp);
+			}
+		}
+		if (!v4_probe_ok)
+			prune_host_list(ap->logopt, &hosts, vers, port);
+	} else {
 		prune_host_list(ap->logopt, &hosts, vers, port);
+	}
 
+dont_probe:
 	if (!hosts) {
 		info(ap->logopt, MODPREFIX "no hosts available");
 		return 1;
--- autofs-5.0.5.orig/modules/replicated.c
+++ autofs-5.0.5/modules/replicated.c
@@ -282,10 +282,10 @@ static unsigned int get_proximity(struct
 	return PROXIMITY_OTHER;
 }
 
-static struct host *new_host(const char *name,
-			     struct sockaddr *addr, size_t addr_len,
-			     unsigned int proximity, unsigned int weight,
-			     unsigned int options)
+struct host *new_host(const char *name,
+		      struct sockaddr *addr, size_t addr_len,
+		      unsigned int proximity, unsigned int weight,
+		      unsigned int options)
 {
 	struct host *new;
 	struct sockaddr *tmp2;
